Add exchange for union with TableScan and Values

rschlussel opened the pull request:
Fixes #10612

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestAddExchangesPlans.java:
```diff
@@ -0,0 +1,99 @@
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+
+public class TestAddExchangesPlans
```
Why not to add to `TestAddExchanges`?

rschlussel replied:
I thought about that, but TestAddExchanges is all about checking the property prioritizing logic (currently there are no plan tests for AddExchanges), so this sort of test seemed out of place. .  I don't feel strongly, though.

rschlussel replied:
I'd prefer to leave it as is or add the tests to TestAddExchanges .  This is specifically mean to be a planner test of the addExchanges rule.  And I'd like to encourage more such tests in the future :).  AddExchanges is super complicated and we don't seem to have planner tests that verify that it's doing what we expect.

rschlussel replied:
Sure.  What would make it clearer?  I want to add planner tests that test that the plan you get for the query after the optimizer runs is what you expect.  We have these kinds of tests for many other optimizers, but not AddExchanges. TestAddExchanges tests the logic of how partitioning preferences work.

rschlussel commented:
I reverted the commit adding an exchange so broadcast joins don't execute on one node because it was causing failures and I didn't finish getting to the bottom of how to fix it.  The other commit (preventing hashjoin/aggregations from happening on one node) is much more important because queries fail without it rather than just being inefficient. 

I'll still look into it, but it was an opportunistic fix, so not sure how quickly.  There are a lot of implicit rules about how exchanges need to work and nothing really enforcing it or even describing it and the code is really complicated.

I also added two small simplifying refactorings here from things I ran into while investigating.

rschlussel commented:
There are tpcds plan changes. could someone test them? q04, q11, q49, q74

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -414,6 +414,24 @@ public PlanWithProperties visitTopN(TopNNode node, PreferredProperties preferred
         {
+            if (isDistributedSortEnabled(session)) {
+                PlanWithProperties child = planChild(node, PreferredProperties.any());
+                if (!child.getProperties().isSingleNode()) {
```
there is a slight difference here where we skip distributed sort if the child is already on a single node.  this is similar to the logic that was happening when this check came after the isSingleNode check below but not exactly the same (the child is different because it was planned with the distributed sort preferred properties and also doesn't depend on whether the data was already sorted -- though I can certainly add that if that's preferred)

findepi replied:
cc @sopel39

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -969,7 +969,7 @@ private Partitioning selectUnionPartitioning(UnionNode node, PreferredProperties
                         .build();
                 PlanWithProperties child = node.getSources().get(sourceIndex).accept(this, childPreferred);
-                if (child.getProperties().isNodePartitionedOn(childPartitioning.getPartitioningColumns(), nullsAndAnyReplicated)) {
+                if (child.getProperties().isNodePartitionedOn(childPartitioning.getPartitioningColumns(), nullsAndAnyReplicated) && !child.getProperties().isSingleNode()) {
```
you could add a comment here

rschlussel replied:
To me it didn't seem worth the added complexity. Also, if all children are all single distributed it doesn't mean they are on the same node.  They could be single distributed on different nodes, and not fit on one node.

rschlussel commented:
@sopel39  can you test the tpcds changes?

sopel39 commented:
@rschlussel our benchmarking cluster is down at the moment

rschlussel commented:
@sopel39  any update on the benchmarking?

sopel39 commented:
@rschlussel we are benchmarking release candidate at the moment. I could run your benchmarks after that

rschlussel commented:
@sopel39 I saw that you ran benchmarks for the semi join PR.  Did you have a chance to benchmark this PR too?

wenleix commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestEliminateSorts.java:
```diff
@@ -76,12 +76,10 @@ public void testNotEliminateSorts()
-                                                        .addFunction(functionCall("row_number", Optional.empty(), ImmutableList.of())),
-                                                anyTree(LINEITEM_TABLESCAN_Q)))));
-
+                                window(windowMatcherBuilder -> windowMatcherBuilder
```
curious: why remove `anyTree`?

rschlussel replied:
anyTree doesn't match zero nodes.  there must have been an intervening node that got removed after the refactor.  I don't recall what it was, but I would guess it's a remote exchange due to the slightly different logic and this test running on a single node.

shixuan-fan replied:
It feels weird when a refactor needs to touch tests. As you mentioned, the logic is different because there are cases that child is single node but does not hit `return` (https://github.com/prestodb/presto/pull/11957/commits/73c23b0d4f325c56093e882a8d972177c3d205d2#diff-3147362e4362824887f1eb890a167d41L428), which might end up having distributed sort enabled. This is not covered after refactoring.

I'm not sure if this would make the plan less optimal or more optimal, but if we could eliminate a remote exchange, maybe the purpose of this commit change from "refactor" to "optimize"?

rschlussel commented:
> "Remove unused argument from selectUnionPartitioning()" looks good.
> 
> Is the right syntax "from selectUnionPartitioning()" or "in selectUnionPartitioning()" ?

Either is fine.  Using "from" specifies where you are removing the thing from (I'm taking the argument out of "selectUnionPartitioning").  Using "in" specifies which particular thing you are removing ("The argument I'm getting rid of is the one in "selectUnionPartitioning").  Practically speaking the two are basically equivalent and either works.

rschlussel commented:
> "Refactor AddExchanges.visitSort()"
> 
> Curious: What's the motivation of this commit?

I was trying to debug an issue with a sort query and it was very confusing because it did all this planning before hitting the if block and then threw it all out because actually distributed sort was enabled.

wenleix commented:
Looks good to me, per offline discussion, let's do the following:

1. Add comment to `TestAddExchangesPlans`.  What you explained in https://github.com/prestodb/presto/pull/11957#discussion_r252041836 is quite good :) 

2. Add a comment about https://github.com/prestodb/presto/pull/11957#discussion_r252040600 (more details added later)

3. Make `visitSort` refactor a separate PR so other engineers can spend a bit time playing with it :) 

4. Request @arhimondr  and @shixuan-fan  to also take a look so we have more engineers have context on optimizers.

sopel39 commented:
@rschlussel here are benchmarks results
![obraz](https://user-images.githubusercontent.com/5989165/52402764-66046a00-2ac5-11e9-881f-a63d19c32465.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402780-6a308780-2ac5-11e9-908f-b5029d10c79d.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402787-6d2b7800-2ac5-11e9-9a95-ab4aad052660.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402797-74528600-2ac5-11e9-8d0d-f28c6c69970a.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402799-79173a00-2ac5-11e9-82ff-01350e4769f4.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402807-7b799400-2ac5-11e9-949d-08ed78ba0543.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402810-7ddbee00-2ac5-11e9-9ae5-3921cbd05881.png)
![obraz](https://user-images.githubusercontent.com/5989165/52402815-803e4800-2ac5-11e9-8c71-6933020a9b42.png)
